### PR TITLE
[Behat] Use Address example factory in Address transformer

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | yes (but only in behats) |
| Related tickets |  |
| License         | MIT |

Starting refactoring of Behat to make code more maintainable.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php [hunk 1]
```diff
@@ -143,8 +140,6 @@ public function create(array $options = [])
             $address->setProvinceName($options['province_name']);
         }
 
-        $options['customer']->addAddress($address);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php]: Removing the feature to assign address to a customer just to make use of this factory in tests is not good... 😕

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php]: Adding an address to the customer in address factory is not good as well, we should set a customer on the address instead.

File: src/Sylius/Behat/Context/Transform/AddressContext.php [hunk 2]
```diff
@@ -99,79 +108,57 @@ public function createEmptyAddress()
...
      * @Transform /^address for "([^"]+)" from "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"$/
-     * @Transform /^"([^"]+)" addressed it to "([^"]+)", "([^"]+)" "([^"]+)" in the "([^"]+)"(?:|, "([^"]+)")$/
-     * @Transform /^of "([^"]+)" in the "([^"]+)", "([^"]+)" "([^"]+)", "([^"]+)"(?:|, "([^"]+)")$/
-     * @Transform /^addressed it to "([^"]+)", "([^"]+)", "([^"]+)" "([^"]+)" in the "([^"]+)"(?:|, "([^"]+)")$/
-     * @Transform /^address (?:|is )"([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"(?:|, "([^"]+)")$/
-     * @Transform /^address as "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"(?:|, "([^"]+)")$/
+     * @Transform /^"([^"]+)" addressed it to "([^"]+)", "([^"]+)" "([^"]+)" in the "([^"]+)", "([^"]+)"$/
+     * @Transform /^of "([^"]+)" in the "([^"]+)", "([^"]+)" "([^"]+)", "([^"]+)", "([^"]+)"$/
+     * @Transform /^addressed it to "([^"]+)", "([^"]+)", "([^"]+)" "([^"]+)" in the "([^"]+)", "([^"]+)"$/
+     * @Transform /^address (?:|is )"([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Transform/AddressContext.php]: This one

File: src/Sylius/Behat/Context/Transform/AddressContext.php [hunk 3]
```diff
@@ -99,79 +108,57 @@ public function createEmptyAddress()
...
+     * @Transform /^address as "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Transform/AddressContext.php]: And this one can be easily merged as `/^address (?:|as |is )"([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"$/`

File: src/Sylius/Behat/Context/Transform/AddressContext.php [hunk 4]
```diff
@@ -99,79 +108,57 @@ public function createEmptyAddress()
...
+        ]);
     }
 
     /**
-     * @Transform /^"([^"]+)" street$/
+     * @Transform /^"([^"]+)" addressed it to "([^"]+)", "([^"]+)" "([^"]+)" in the "([^"]+)"$/
+     * @Transform /^of "([^"]+)" in the "([^"]+)", "([^"]+)" "([^"]+)", "([^"]+)"$/
+     * @Transform /^addressed it to "([^"]+)", "([^"]+)", "([^"]+)" "([^"]+)" in the "([^"]+)"$/
+     * @Transform /^address (?:|is )"([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"$/
+     * @Transform /^address as "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)"$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Transform/AddressContext.php]: Same merge as mentioned above.

File: src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php [hunk 5]
```diff
@@ -175,11 +163,41 @@ private function assertProvinceCodeIsValid($provinceCode, $countryCode)
...
+        if ($country->hasProvinces()) {
+            return $this->getProvinceCode($country->getProvinces(), $options['province_name']);
+        }
+
+        return $options['province_name'];
+    }
+
+    /**
+     * @param Collection $provinces
+     * @param $provinceName
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php]: Missing typehint.

File: src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php [hunk 6]
```diff
@@ -175,11 +163,41 @@ private function assertProvinceCodeIsValid($provinceCode, $countryCode)
     }
 
     /**
-     * @param string $countryCode
+     * @param array $options
+     *
+     * @return string
+     *
+     * @throws \InvalidArgumentException
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php]: In fact it does not throw anything, it just don't catch it.

File: features/user/managing_customers/seeing_province_created_manually_on_customer_details_page.feature [hunk 7]
```diff
@@ -6,7 +6,8 @@ Feature: Seeing a province on customer's details
 
     @ui
     Scenario: Seeing customer's addresses
-        Given I am logged in as an administrator
+        Given the store operates in "United Kingdom"
+        And I am logged in as an administrator
```
Comment (michalmarcinkowski) [inline features/user/managing_customers/seeing_province_created_manually_on_customer_details_page.feature]: Should be the last `Given` step.

File: src/Sylius/Behat/Context/Transform/AddressContext.php [hunk 8]
```diff
@@ -99,79 +109,59 @@ public function createEmptyAddress()
...
         list($firstName, $lastName) = explode(' ', $name);
 
-        return $this->createAddress($countryCode, $firstName, $lastName, $city, $street, $postcode, $provinceName);
+        return $this->exampleAddressFactory->create([
+            'country_code' => $this->countryNameConverter->convertToCode($countryName),
+            'first_name' => $firstName,
+            'last_name' => $lastName,
+            'company' => null,
+            'customer' => null,
+            'phone_number' => null,
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Transform/AddressContext.php]: Do we need to pass `null` for optional fields like `phone_number` or `company`? Can't we just skip them?

File: src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php [hunk 9]
```diff
@@ -135,15 +137,11 @@ public function create(array $options = [])
...
-            $address->setProvinceCode($options['province_code']);
-        } else {
-            $this->assertCountryHasNoProvinces($options['country_code']);
-            $address->setProvinceName($options['province_name']);
-        }
+        $this->resolveCountryProvince($options, $address);
 
-        $options['customer']->addAddress($address);
+        if (isset($options['customer'])) {
+            $options['customer']->addAddress($address);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Fixture/Factory/AddressExampleFactory.php]: We allow for `null` value, shouldn't we verify we don't call this method on `null` object?

**Comment by michalmarcinkowski** [comment]
Thanks Łukasz!

**Pull request merged**